src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/BookingHistoryService.kt

```diff
@@ -0,0 +1,79 @@
+package uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.service
+
+import jakarta.persistence.EntityNotFoundException
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.entity.BookingHistory
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.entity.BookingHistoryAppointment
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.entity.VideoBooking
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.request.PrisonerDetails
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.repository.BookingHistoryRepository
+
+@Service
+class BookingHistoryService(private val bookingHistoryRepository: BookingHistoryRepository) {
+
+  fun getHistoryByVideoBookingId(videoBookingId: Long): List<BookingHistory> =
+    bookingHistoryRepository.findAllByVideoBookingIdOrderByCreatedTime(videoBookingId)
+
+  fun getHistoryByBookingHistoryId(bookingHistoryId: Long): BookingHistory =
+    bookingHistoryRepository.findById(bookingHistoryId)
+      .orElseThrow { EntityNotFoundException("Video booking history with ID $bookingHistoryId not found") }
+
+  @Transactional
+  fun createBookingHistoryForCourt(historyType: String, booking: VideoBooking, prisonerDetails: List<PrisonerDetails>) =
```

Comment:
If I understand correctly can you not just pull of the appointments from the booking directly instead of passing in the prisoner details list?  I might be misunderstanding but the history is created after a booking has been created/changed?

Reply:
I could be wrong but something like this:

```
  @Transactional
  fun createBookingHistoryForCourt(historyType: String, booking: VideoBooking) =
    BookingHistory(
      videoBookingId = booking.videoBookingId,
      historyType = historyType,
      courtId = booking.court?.courtId,
      hearingType = booking.hearingType,
      videoUrl = booking.videoUrl,
      comments = booking.comments,
      createdBy = booking.createdBy,
    ).apply {
      addBookingHistoryAppointments(getAppointmentsForHistory(this, booking))
    }.also { history ->
      bookingHistoryRepository.saveAndFlush(history)
    }

  @Transactional
  fun createBookingHistoryForProbation(historyType: String, booking: VideoBooking) =
    BookingHistory(
      videoBookingId = booking.videoBookingId,
      historyType = historyType,
      probationTeamId = booking.probationTeam?.probationTeamId,
      probationMeetingType = booking.probationMeetingType,
      videoUrl = booking.videoUrl,
      comments = booking.comments,
      createdBy = booking.createdBy,
    ).apply {
      addBookingHistoryAppointments(getAppointmentsForHistory(this, booking))
    }.also { history ->
      bookingHistoryRepository.saveAndFlush(history)
    }

  private fun getAppointmentsForHistory(history: BookingHistory, booking: VideoBooking) =
    booking.appointments().map { app ->
      BookingHistoryAppointment(
        bookingHistory = history,
        prisonCode = app.prisonCode,
        prisonerNumber = app.prisonerNumber,
        appointmentDate = app.appointmentDate,
        appointmentType = app.appointmentType,
        prisonLocKey = app.prisonLocKey,
        startTime = app.startTime,
        endTime = app.endTime,
      )
    }

```

Reply:
Yes - can now, since your recent change to add appointments to the booking :-) I'll sort this now.

Reply:
And done, as suggested.